backend/lib/plugins/socket.js

```diff
@@ -0,0 +1,555 @@
+
+    const threadsRich = []; // Rich because will have lastMessage, and topic
+    for (let i = 0; i < threads.length; i++) {
+      const thread = threads[i].toObject();
+      const [lastMessageErr, lastMessage] = await this.to(
+        Message.findOne({ threadId: thread._id }).sort({ createdAt: -1 }),
+      );
+      const [lastEmbedMessageErr, lastEmbedMessage] = await this.to(
+        Message.findOne({ threadId: thread._id, postRef: { $ne: null } }).sort({
+          createdAt: -1,
+        }),
+      );
+
+      if (lastMessage) thread.lastMessage = lastMessage;
+      if (lastEmbedMessage) thread.topic = lastEmbedMessage.postRef.title;
+      threadsRich.push(thread);
+    }
+    res({ code: 200, data: threadsRich });
+  });
+
+  socket.on("GET_USER_STATUS", async (id, res) => {
+    const status = (await getSocketIdByUserId(this, id)) ? "online" : "offline";
+    res({ code: 200, data: status });
+  });
+
+  socket.on("GET_CHAT_LOG", async (data, res) => {
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
+    if (!userInRoom) return res({ code: 401, message: "Unauthorized" });
+
+    const [messagesErr, messages] = await this.to(
+      Message.find({ threadId: data.threadId })
+        .limit(20)
+        .sort({ createdAt: -1 }),
+    );
+
+    if (messagesErr)
+      return res({ code: 500, message: "Internal server error" });
+    res({ code: 200, data: messages });
+  });
+
+  socket.on("GET_CHAT_LOG_MORE", async (data, res) => {
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
+    if (!userInRoom) return res({ code: 401, message: "Unauthorized" });
+
+    const [messagesErr, messages] = await this.to(
+      Message.find({ threadId: data.threadId })
+        .limit(20)
+        .skip(data.skip)
+        .sort({ createdAt: -1 }),
+    );
+
+    if (messagesErr)
+      return res({ code: 500, message: "Internal server error" });
+    res({ code: 200, data: messages });
+  });
+
+  socket.on("SEND_MESSAGE", async (data, res) => {
+    userId = socket.userId;
+    let userInRoom = await isUserInRoom(this, data.threadId, socket.id);
+    if (!data.threadId && !userInRoom)
+      return res({ code: 401, message: "Unauthorized" });
+    var [threadErr, thread] = await this.to(
+      Thread.findOne({
+        _id: data.threadId,
+        "participants.id": userId,
+      }),
+    );
+    if (threadErr) return res({ code: 500, message: "Internal server error" });
+    if (!thread) return res({ code: 401, message: "Unauthorized" });
+
+    // someone blocked someone else, no one can send a message
+    if (thread.participants.find((p) => p.status == "blocked"))
+      return res({ code: 401, message: "Unauthorized" });
+    // sender is still "pending", they can't send until they accept
+    if (
+      thread.participants.find((p) => p.id == userId && p.status == "pending")
+    )
+      return res({ code: 401, message: "Unauthorized" });
```

Comment:
I know this is valid syntax but the formatting looks awkward, so I would add the braces. Also `===` rather than `==` since type coercion is janky 😄 
```suggestion
    if (
      thread.participants.find((p) => p.id === userId && p.status === "pending")
    ) {
      return res({ code: 401, message: "Unauthorized" });
    }
```